python/pyxet/pyxet/cli.py

```diff
@@ -152,69 +158,96 @@ def _copy(source, destination, recursive = True, _src_fs=None, _dest_fs=None):
+
                 relpath = _ltrim_match(path, src_root_dir).lstrip('/')
                 if dest_path == '/':
                     dest_for_this_path = f"/{relpath}"
                 else:
                     dest_for_this_path = f"{dest_path}/{relpath}"
                 dest_dir = '/'.join(dest_for_this_path.split('/')[:-1])
-                dest_fs.makedirs(dest_dir, exist_ok=True)
+                dest_directories.append((dest_fs, dest_dir))
 
                 futures.append(
                     executor.submit(
-                        _copy,
+                        _build_source_destination_list,
                         f"{src_fs.protocol}://{path}",
                         f"{dest_fs.protocol}://{dest_for_this_path}",
                         recursive=True,
                         _src_fs=src_fs,
                         _dest_fs=dest_fs,
                     )
                 )
+
+            cp_commands = []
             for future in futures:
-                future.result()
-        return
+                _cp_commands, _dest_directories = future.result()
+                cp_commands += _cp_commands
+                dest_directories += _dest_directories
+
+            return cp_commands, dest_directories 
 
     # Handling directories
     if src_isdir:
-        # Recursively copy
-        # xet cp_file can cp directories
-        if srcproto == 'xet' and destproto == 'xet':
-            print(f"Copying {src_path} to {dest_path}...")
-            dest_fs.cp_file(src_path, dest_path)
-            return
-        with ThreadPoolExecutor() as executor:
-            futures = []
-            for path,info in src_fs.find(src_path, detail=True).items():
-                if info['type'] == 'directory' and not recursive:
-                    continue
-                # Note that path is a full path
-                # we need to relativize to make the destination path
-                relpath = _ltrim_match(path, src_path).lstrip('/')
-                if dest_path == '/':
-                    dest_for_this_path = f"/{relpath}"
-                else:
-                    dest_for_this_path = f"{dest_path}/{relpath}"
-                dest_dir = '/'.join(dest_for_this_path.split('/')[:-1])
-                dest_fs.makedirs(dest_dir, exist_ok=True)
-                # Submitting copy jobs to thread pool
-                futures.append(
-                        executor.submit(
-                            _single_file_copy,
-                            src_fs,
-                            f"{path}",
-                            dest_fs,
-                            dest_for_this_path))
-            # Waiting for all copy jobs to complete
-            for future in futures:
-                future.result()
-        return
+        cp_commands = []
+        dest_directories = []
+        for path,info in src_fs.find(src_path, detail=True).items():
+            if info['type'] == 'directory': # glob will hit within this directory
```

Comment:
I think we should have `and not recursive` here because of cp like behavior?

Reply:
This one is tricky and needs a rewrite to match cp behavior.  src_fs.find(...) is itself recursive, which may not be what we want here.  Will work on this.